File: backend/domain/value_objects/slash_commands.py
"""
Slash command parser for chat mode.

Parses player input for /chat and /end commands to control chat mode.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Optional


class SlashCommandType(Enum):
    """Types of slash commands supported."""

    CHAT = "chat"  # Enter chat mode
    END = "end"  # Exit chat mode
    NONE = "none"  # Not a slash command


@dataclass
class ParsedCommand:
    """Result of parsing a slash command."""

    command_type: SlashCommandType
    args: Optional[str] = None


def parse_slash_command(action_text: str) -> ParsedCommand:
    """
    Parse action text for slash commands.

    Args:
        action_text: The player's input text

    Returns:
        ParsedCommand with the command type and any arguments

    Examples:
        >>> parse_slash_command("/chat")
        ParsedCommand(command_type=SlashCommandType.CHAT, args=None)
        >>> parse_slash_command("/end")
        ParsedCommand(command_type=SlashCommandType.END, args=None)
        >>> parse_slash_command("hello world")
        ParsedCommand(command_type=SlashCommandType.NONE, args=None)
    """
    text = action_text.strip()

    # Check for /chat command
    if text.lower() == "/chat":
        return ParsedCommand(SlashCommandType.CHAT)

    # Check for /end command
    if text.lower() == "/end":
        return ParsedCommand(SlashCommandType.END)

    # Not a recognized slash command
    return ParsedCommand(SlashCommandType.NONE)

File: backend-ts/src/domain/slash-commands.ts
/**
 * Chat-mode slash commands. Port of
 * `backend/domain/value_objects/slash_commands.py`.
 *
 * `/chat` drops the player into free-form conversation with the NPCs at their
 * location; `/end` leaves it and hands the transcript to the Chat Summarizer.
 * Everything else is a normal action and goes to the Action Manager.
 */

/**
 * Python's `SlashCommandType` is a bare `Enum`, not a `str, Enum` — it is only
 * ever compared, never serialised — so nothing depends on these exact strings
 * crossing a wire. They are kept anyway so a log line reads the same.
 */
export const SLASH_COMMAND_TYPES = ['chat', 'end', 'none'] as const

export type SlashCommandType = (typeof SLASH_COMMAND_TYPES)[number]

export interface ParsedCommand {
  commandType: SlashCommandType
  /**
   * Always `null` today. Python declares `args: Optional[str] = None` and
   * never populates it; the field is kept so adding an argument-taking command
   * does not change every call site's shape.
   */
  args: string | null
}

/**
 * Classify a line of player input.
 *
 * The match is **exact** on the trimmed, lowercased text — not a prefix test.
 * `/chatter with the innkeeper` is an ordinary action and must reach the
 * Action Manager; a `startsWith` here would silently swallow it into chat mode
 * instead. That is the whole reason this function exists rather than being
 * inlined as a comparison.
 */
export function parseSlashCommand(actionText: string): ParsedCommand {
  const text = actionText.trim().toLowerCase()

  if (text === '/chat') return { commandType: 'chat', args: null }
  if (text === '/end') return { commandType: 'end', args: null }

  return { commandType: 'none', args: null }
}
